xls/integrator/ir_integrator.h

```diff
@@ -0,0 +1,63 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef XLS_INTEGRATOR_IR_INTEGRATOR_H_
+#define XLS_INTEGRATOR_IR_INTEGRATOR_H_
+
+#include "absl/status/statusor.h"
+#include "xls/ir/function.h"
+#include "xls/ir/package.h"
+
+namespace xls {
+
+// Class used to integrate separate functions into a combined, reprogrammable
+// circuit that can be configured to have the same functionality as the
+// input functions. The builder will attempt to construct the integrated
+// function such that hardware common to the input functions is consolidated.
+// Note that this is distinct from function inlining. With inlining, a function
+// call is replaced by the body of the function that is called.  With function
+// integration, we take separate functions that do not call each other and
+// combine the hardware used to implement the functions.
+class IntegrationBuilder {
+ public:
+  IntegrationBuilder(absl::Span<const Function* const> input_functions) {
```

Comment:
How about making the constructor private and just having a static build method. It could return a std::unique_ptr<Package> and a Function pointer. This avoids potential ownership issues where you require the original functions to be kept around until Build is called.

Reply:
Done.  However, absl::make_unique needs access to a public constructor to work.  I implemented a hacky workaround where we call make_unqiue for a private sublcass with a public constructor, similar to the one presented here:
https://seanmiddleditch.com/enabling-make-unique-with-private-constructors/